src/Apropos/Gen.hs

```diff
@@ -22,17 +29,130 @@ module Apropos.Gen (
+    Right a -> pure a
+
+forAll :: Show a => Gen a -> PropertyT IO a
+forAll g = do
+  (ee,labels) <- H.forAll $ runWriterT (runExceptT $ gen g)
+  mapM_ (H.label . fromString) labels
+  case ee of
+    Left Retry -> H.footnote "retry limit reached" >> H.failure
+    Left (GenException err) -> H.footnote err >> H.failure
+    Right a -> pure a
+
+data Morph m where
+  Source :: Gen m -> Morph m
+  Morph :: Morph m -> (m -> Gen [m -> Gen m]) -> Morph m
+
+--TODO use Morphism instead then we can print the Morphism name
+--i.e. Morph :: Morph p m -> (m -> Gen [Morphism p m]) -> Morph p m
+--also Source has been introduced - we want to be able to partition the space with a set of sources! XD
+instance Show (m -> Gen m) where
+  show _ = "GenMorphism"
+
+morphAsGen :: Morph m -> Gen m
+morphAsGen (Source g) = g
+morphAsGen (Morph m g) = do
+  n <- morphAsGen m
+  t <- g n
+  foldl (>=>) pure t n
+
+morphContain :: Show m => Morph m -> PropertyT IO m
+morphContain (Source g) = forAll g
+morphContain (Morph m g) = do
+  n <- morphContain m
+  t <- forAllNoShrink $ g n
+  foldl (\a b -> a >>= forAll . b) (pure n) t
+
+morphContainRetry :: Show m => Int -> Morph m -> PropertyT IO m
+morphContainRetry retries m = do
+  r <- morphContainRetry' retries m
+  case r of
+    Nothing -> H.footnote ("retry limit (" <> show retries <> ") reached") >> H.failure
+    Just so -> pure so
+
+-- hmmmmm - let's refactor Morph
+morphContainRetry' :: Show m => Int -> Morph m -> PropertyT IO (Maybe m)
+morphContainRetry' _ (Source g) = forAllRetryToMaybe g
+morphContainRetry' retries (Morph m g) = do
+  handleRetries' retries (morphContainRetry' retries m) $ \a -> do
+    t <- forAllNoShrink $ g a
+    let c = steps' retries t
+    c a
+
+-- naming here is not great
+steps :: Int -> [a -> PropertyT IO (Maybe a)] -> (a -> PropertyT IO (Maybe a))
+steps _ [] = pure . Just
+steps retries (c:cs) = \x -> handleRetries' retries (c x) (steps retries cs)
+
+-- TODO we want to pass in a scale parameter that scales each step on retry
+steps' :: Show a => Int -> [a -> Gen a] -> (a -> PropertyT IO (Maybe a))
+steps' retries gs = do
+  let rs = (forAllRetryToMaybe .) <$> gs
+  steps retries rs
+
+forAllRetryToMaybe :: Show a => Gen a -> PropertyT IO (Maybe a)
+forAllRetryToMaybe g = do
+  (ee,labels) <- H.forAll $ runWriterT (runExceptT $ gen g)
+  mapM_ (H.label . fromString) labels
+  case ee of
+    Left Retry -> pure Nothing
+    Left (GenException err) -> H.footnote err >> H.failure
+    Right a -> pure $ Just a
+
+
+handleRetries' :: forall m a b. Monad m => Int -> m (Maybe a) -> (a -> m (Maybe b)) -> m (Maybe b)
+handleRetries' n g c = go n
+  where
+    go :: Int -> m (Maybe b)
+    go l = do
+      res <- g
+      case res of
+        Just r -> do
+          res2 <- c r
+          case res2 of
+            Nothing -> go (l - 1)
+            Just so -> pure $ Just so
+        Nothing ->
+          if l < 1
+             then pure Nothing
+             else go (l - 1)
```

Comment:
This does the backtracking retries with each step contained in PropertyT so that the shrink trees are small.
TODO: when we retry we want to scale up the size of the Gen computation like we do in `genRetry`.